### PR TITLE
Add support for distributed CUDA mapreduce

### PR DESCRIPTION
**Pull request by sriharshakandala**
- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: examples/hybrid/box/bubble_3d_invariant_rhoe.jl [hunk 1]
```diff
@@ -370,11 +370,13 @@ function bubble_3d_invariant_ρe(ARGS, comms_ctx, ::Type{FT}) where {FT}
...
 
-    if ClimaComms.iamroot(comms_ctx)
-        @info "Context information" device = comms_ctx.device context =
-            comms_ctx nprocs = ClimaComms.nprocs(comms_ctx) Float_type = FT resolution =
-            resolution
+    logger_stream = ClimaComms.iamroot(comms_ctx) ? stderr : devnull
+    prev_logger = global_logger(ConsoleLogger(logger_stream, Logging.Info))
+    atexit() do
+        global_logger(prev_logger)
     end
```
Comment (simonbyrne) [inline examples/hybrid/box/bubble_3d_invariant_rhoe.jl:377]: I think we can get rid of this
```suggestion
```

**Review by simonbyrne — COMMENTED** [review]
Looks good, but I think it needs some unit tests.

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

File: test/Fields/reduction_cuda_distributed.jl [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    ne = 72
+    Nq = 4
+    ndof = ne * ne * 6 * Nq * Nq
+    @info "Configuration" device = context.device nprocs =
+        ClimaComms.nprocs(context) Ne = ne Nq = Nq ndof = ndof FT = FT
+    R = FT(6.37122e6) # radius of earth
+    domain = Domains.SphereDomain(R)
+    mesh = Meshes.EquiangularCubedSphere(domain, ne)
+    quad = Spaces.Quadratures.GLL{Nq}()
+    grid_topology = Topologies.Topology2D(context, mesh)
```
Comment (akshaysridhar) [inline test/Fields/reduction_cuda_distributed.jl:39]: Tests define different topologies depending on context, then check that functions return approx identical values on cpu / gpu devices; seems reasonable to me!

Are tests for `spacefillingcurve` implicit in this setup? For instance, in ClimaAtmos the topology is constructed with 

```
 topology = Topologies.Topology2D(
            comms_ctx,
            mesh,
            Topologies.spacefillingcurve(mesh),
        )
```

Comment (akshaysridhar) [reply test/Fields/reduction_cuda_distributed.jl:39]: Resolving. This is considered in `118` below.

Comment (sriharshakandala) [reply test/Fields/reduction_cuda_distributed.jl:39]: For `ClimaCore.jl`, `linearindices` are the default. 
https://github.com/CliMA/ClimaCore.jl/blob/main/src/Topologies/topology2d.jl#L187

Comment (sriharshakandala) [reply test/Fields/reduction_cuda_distributed.jl:39]: The results of the tests should be independent of the partitioning strategy. Some of the tests compare reduced values with analytical estimates.

File: examples/hybrid/box/bubble_3d_invariant_rhoe.jl
Comment (akshaysridhar) [inline examples/hybrid/box/bubble_3d_invariant_rhoe.jl:1]: It seems there is an issue here - executing this file on the PR branch commit `0f2300c` results in the following error message: 
```
julia> include("bubble_3d_invariant_rhoe.jl")
┌ Info: Context information
│   device = ClimaComms.CUDADevice()
│   context = ClimaComms.MPICommsContext{ClimaComms.CUDADevice, MPI.Comm}(ClimaComms.CUDADevice(), MPI.Comm(Ptr{Nothing} @0x00007fa160a30ef0))
│   nprocs = 1
│   Float_type = Float64
└   resolution = "low"
┌ Info: Initial condition
│   energy_0 = 2.301635903034133e13
│   mass_0 = 1.1147514015211978e9
│   κ₄ = 193.32836229901798
└   ClimaComms.mypid(comms_ctx) = 1
ERROR: LoadError: UndefVarError: fw not defined
Stacktrace:
 [1] bubble_3d_invariant_ρe(ARGS::Vector{String}, comms_ctx::ClimaComms.MPICommsContext{ClimaComms.CUDADevice, MPI.Comm}, #unused#::Type{Float64})
   @ Main /central/home/asridhar/ClimaCore.jl/examples/hybrid/box/bubble_3d_invariant_rhoe.jl:418
 [2] top-level scope
   @ /central/home/asridhar/ClimaCore.jl/examples/hybrid/box/bubble_3d_invariant_rhoe.jl:544
 [3] include(fname::String)
   @ Base.MainInclude ./client.jl:476
 [4] top-level scope
   @ REPL[1]:1
in expression starting at /central/home/asridhar/ClimaCore.jl/examples/hybrid/box/bubble_3d_invariant_rhoe.jl:544
```

File: test/Fields/reduction_cuda_distributed.jl (hunk 2 above)
Comment (akshaysridhar) [reply test/Fields/reduction_cuda_distributed.jl:39]: Noted thank you!

File: examples/hybrid/box/bubble_3d_invariant_rhoe.jl
Comment (sriharshakandala) [reply examples/hybrid/box/bubble_3d_invariant_rhoe.jl:1]: This has been fixed in the latest commit!

Comment (akshaysridhar) [reply examples/hybrid/box/bubble_3d_invariant_rhoe.jl:1]: Line 432 : `fw` -> `w` ?

Comment (akshaysridhar) [inline examples/hybrid/box/bubble_3d_invariant_rhoe.jl:1]: On commit `b6e2e520`
```
julia> include("bubble_3d_invariant_rhoe.jl")
┌ Info: Context information
│   device = ClimaComms.CUDADevice()
│   context = ClimaComms.MPICommsContext{ClimaComms.CUDADevice, MPI.Comm}(ClimaComms.CUDADevice(), MPI.Comm(Ptr{Nothing} @0x00007f4501072ef0))
│   nprocs = 1
│   Float_type = Float64
└   resolution = "low"
┌ Info: Initial condition
│   energy_0 = 2.301635903034133e13
│   mass_0 = 1.1147514015211978e9
│   κ₄ = 193.32836229901798
└   ClimaComms.mypid(comms_ctx) = 1
Walltime = 106.846003328 seconds
┌ Info: summary
│   Es[end] = 2.3016359030341336e13
│   Mass[end] = 1.1147514015211985e9
└   ClimaComms.mypid(comms_ctx) = 1
┌ Warning: No strict ticks found
└ @ PlotUtils ~/.julia/packages/PlotUtils/mHQ0Q/src/ticks.jl:191
┌ Warning: No strict ticks found
└ @ PlotUtils ~/.julia/packages/PlotUtils/mHQ0Q/src/ticks.jl:191
retcode: Success
Interpolation: 1st order linear
t: 71-element Vector{Float64}: ...
u: u: 71-element Vector ...
ERROR: Scalar indexing is disallowed.
```
Running on `Tesla P100-PCIE-16GB` with `CUDA driver 12.1
NVIDIA driver 510.47.3, originally for CUDA 11.6`

Comment (akshaysridhar) [reply examples/hybrid/box/bubble_3d_invariant_rhoe.jl:1]: Relative change in energy is `3.394324875494498e-16`, and in mass is `6.416280224708795e-16`. These are reasonable ; seems like we need to modify something in the post-proc script (?)

**Comment by akshaysridhar** [comment]
Buildkite check seems to run to completion but no artifacts are generated - is this expected ? 
```


[Image artifact 'hybrid/box/output/bubble_3d_invariant_rhoe/energy_low_resolution.png' not found]
--
Failed
 to create 
'/central/scratch/esm/slurm-buildkite/climacore-ci/2263/climacore-ci/output/gpu_rising_bubble_3d_hybrid_invariant_rhoe/report.nsys-rep':
 No such file or directory.
 ```

File: examples/hybrid/box/bubble_3d_invariant_rhoe.jl
Comment (akshaysridhar) [reply examples/hybrid/box/bubble_3d_invariant_rhoe.jl:1]: Above error occurs when launching test case in REPL; job runs to completion when launched directly with terminal command line options.

**Comment by sriharshakandala** [comment]
bors try

**Review by akshaysridhar — APPROVED** [review]

File: examples/hybrid/box/bubble_3d_invariant_rhoe.jl
Comment (sriharshakandala) [reply examples/hybrid/box/bubble_3d_invariant_rhoe.jl:1]:
```
[skandala@hpc-24-38 examples]$ julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.5 (2023-01-08)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> include("hybrid/box/bubble_3d_invariant_rhoe.jl");
┌ Info: Context information
│   device = ClimaComms.CUDADevice()
│   context = ClimaComms.SingletonCommsContext{ClimaComms.CUDADevice}(ClimaComms.CUDADevice())
│   nprocs = 1
│   Float_type = Float64
└   resolution = "low"
┌ Info: Initial condition
│   energy_0 = 2.301635903034133e13
│   mass_0 = 1.1147514015211978e9
│   κ₄ = 193.32836229901798
└   ClimaComms.mypid(comms_ctx) = 1
Walltime = 96.793503761 seconds
┌ Info: summary
│   Es[end] = 2.3016359030341336e13
│   Mass[end] = 1.1147514015211985e9
└   ClimaComms.mypid(comms_ctx) = 1
┌ Warning: No strict ticks found
└ @ PlotUtils ~/.julia/packages/PlotUtils/mHQ0Q/src/ticks.jl:191
┌ Warning: No strict ticks found
└ @ PlotUtils ~/.julia/packages/PlotUtils/mHQ0Q/src/ticks.jl:191

julia> 
```

File: src/Fields/mapreduce_cuda.jl [hunk 3]
```diff
@@ -36,9 +61,9 @@ function LinearAlgebra.norm(
     if p == 2
         # currently only one which supports structured types
```
Comment (simonbyrne) [inline src/Fields/mapreduce_cuda.jl:62]:
```suggestion
        # currently only one which supports structured types
        # TODO: perform map without allocation new field
```

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try-

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
> Buildkite check seems to run to completion but no artifacts are generated - is this expected ?
> 
> ```
> 
> 
> [Image artifact 'hybrid/box/output/bubble_3d_invariant_rhoe/energy_low_resolution.png' not found]
> --
> Failed
>  to create 
> '/central/scratch/esm/slurm-buildkite/climacore-ci/2263/climacore-ci/output/gpu_rising_bubble_3d_hybrid_invariant_rhoe/report.nsys-rep':
>  No such file or directory.
> ```

This is now fixed!

File: src/Fields/mapreduce_cuda.jl [hunk 4]
```diff
@@ -35,10 +60,11 @@ function LinearAlgebra.norm(
 )
     if p == 2
         # currently only one which supports structured types
+        # TODO: perform map without allocation new field
         if normalize
-            sqrt.(Statistics.mean(LinearAlgebra.norm_sqr, field))
+            sqrt.(Statistics.mean(LinearAlgebra.norm_sqr.(field)))
```
Comment (charleskawczynski) [inline src/Fields/mapreduce_cuda.jl:65]: Can we do this without allocating intermediate fields?

File: src/Fields/mapreduce_cuda.jl [hunk 5]
```diff
@@ -35,10 +60,11 @@ function LinearAlgebra.norm(
...
         else
-            sqrt.(sum(LinearAlgebra.norm_sqr, field))
+            sqrt.(sum(LinearAlgebra.norm_sqr.(field)))
```
Comment (charleskawczynski) [inline src/Fields/mapreduce_cuda.jl:67]: Can we do this without allocating intermediate fields?

**Review by simonbyrne — APPROVED** [review]

**Pull request merged**